### PR TITLE
Support for viewing keychain items

### PR DESCRIPTION
**Pull request by henrayluo**
I think keychain data is useful as NSUserDefaults.

### CODE REVIEW

**Comment by NSExceptional** [comment]
This looks awesome!

Do you mind if I make a few minor style changes before I merge this? Or I can just tell you what I'd like changed, but it's all very nicpicky.

**Comment by henrayluo** [comment]
> This looks awesome!
> 
> Do you mind if I make a few minor style changes before I merge this? Or I can just tell you what I'd like changed, but it's all very nicpicky.

Sure, Change whatever you like. Thank you.

File: Classes/GlobalStateExplorers/Globals/FLEXGlobalsTableViewController.m [hunk 1]
```diff
@@ -56,6 +57,8 @@ + (FLEXGlobalsEntry *)globalsEntryForRow:(FLEXGlobalsRow)row
     switch (row) {
         case FLEXGlobalsRowAppClasses:
             return [FLEXClassesTableViewController flex_concreteGlobalsEntry:row];
+        case FLEXGlobalsRowAppKeyChainItems:
+            return [FLEXKeyChainTableViewController flex_concreteGlobalsEntry:row]; // ray add
```
Comment (matrush) [inline Classes/GlobalStateExplorers/Globals/FLEXGlobalsTableViewController.m]:
```suggestion
            return [FLEXKeyChainTableViewController flex_concreteGlobalsEntry:row];
```

File: Classes/GlobalStateExplorers/Globals/FLEXGlobalsTableViewController.m [hunk 2]
```diff
@@ -121,6 +124,7 @@ + (FLEXGlobalsEntry *)globalsEntryForRow:(FLEXGlobalsRow)row
             @[ // FLEXGlobalsSectionAppShortcuts
                 [self globalsEntryForRow:FLEXGlobalsRowMainBundle],
                 [self globalsEntryForRow:FLEXGlobalsRowUserDefaults],
+                [self globalsEntryForRow:FLEXGlobalsRowAppKeyChainItems], //ray add
```
Comment (matrush) [inline Classes/GlobalStateExplorers/Globals/FLEXGlobalsTableViewController.m]:
```suggestion
                [self globalsEntryForRow:FLEXGlobalsRowAppKeyChainItems],
```

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
+//
+//  FLEXKeyChainTableViewController.h
+//  FLEX
+//
+//  Created by ray on 2019/8/17.
+//  Copyright © 2019 Flipboard. All rights reserved.
+//
+
+#import "FLEXTableViewController.h"
+#import "FLEXGlobalsEntry.h"
```
Comment (matrush) [inline Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.h]: These imports should be sorted by alphabetic order

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+//
+//  Created by ray on 2019/8/17.
+//  Copyright © 2019 Flipboard. All rights reserved.
+//
+
+#import "FLEXKeyChainTableViewController.h"
+#import "FLEXObjectExplorerFactory.h"
+#import "FLEXUtility.h"
+#import "FLEXKeychain.h"
+#import "FLEXKeychainQuery.h"
```
Comment (matrush) [inline Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m]: Same, please re-order these imports

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    [self.tableView reloadData];
+}
+
+#pragma mark - FLEXGlobalsEntry
+
++ (NSString *)globalsEntryTitle:(FLEXGlobalsRow)row {
+    return [NSString stringWithFormat:@"🔑  %@ KeyChain", [FLEXUtility applicationName]];
+}
+
++ (UIViewController *)globalsEntryViewController:(FLEXGlobalsRow)row {
```
Comment (matrush) [inline Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m]:
```suggestion
+ (UIViewController *)globalsEntryViewController:(FLEXGlobalsRow)row
{
```

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        }
+    }
+    
+    _keyChainItems = [FLEXKeychain allAccounts];
+    [self.tableView reloadData];
+}
+
+#pragma mark - FLEXGlobalsEntry
+
++ (NSString *)globalsEntryTitle:(FLEXGlobalsRow)row {
```
Comment (matrush) [inline Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m]:
```suggestion
+ (NSString *)globalsEntryTitle:(FLEXGlobalsRow)row
{
```

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    
+    UIBarButtonItem *barbtn = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemTrash target:self action:@selector(clearKeyChain)];
+    self.navigationItem.rightBarButtonItem = barbtn;
+
+    
+    _keyChainItems = [FLEXKeychain allAccounts];
+    self.title = [NSString stringWithFormat:@"🔑 KeyChains (%lu)", (unsigned long)self.keyChainItems.count];
+}
+
+- (void)clearKeyChain {
```
Comment (matrush) [inline Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m]:
```suggestion
- (void)clearKeyChain
{
```

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+@interface FLEXKeyChainTableViewController ()
+
+@property (nonatomic) NSArray<NSDictionary *> *keyChainItems;
+
+@end
+
+@implementation FLEXKeyChainTableViewController
+
+- (void)viewDidLoad {
```
Comment (matrush) [inline Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m]:
```suggestion
- (void)viewDidLoad
{
```

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    [super viewDidLoad];
+    
+    UIBarButtonItem *barbtn = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemTrash target:self action:@selector(clearKeyChain)];
```
Comment (matrush) [inline Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m]: This variable is not needed, you can create it inline

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.h [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+#else
+	#import <Foundation/Foundation.h>
+#endif
+
+/**
+ Error code specific to SSKeychain that can be returned in NSError objects.
+ For codes returned by the operating system, refer to SecBase.h for your
+ platform.
+ */
+typedef NS_ENUM(OSStatus, SSKeychainErrorCode) {
```
Comment (matrush) [inline Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.h]: What does `SS` means here? Prefer to use `FLEXKeychain` as prefix

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.m [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+//  SSKeychain
+//
+//  Created by Sam Soffes on 5/19/10.
+//  Copyright (c) 2010-2014 Sam Soffes. All rights reserved.
+//
+
+#import "FLEXKeychain.h"
+#import "FLEXKeychainQuery.h"
+
+NSString *const kSSKeychainErrorDomain = @"com.samsoffes.sskeychain";
```
Comment (matrush) [inline Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.m]: It looks like this is from
https://github.com/samsoffes/sskeychain/blob/master/SSKeychain.m
have you checked any copyright policy?

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.h (hunk 3 above)
Comment (henrayluo) [reply Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.h]: Ok, i will change it.

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m (hunk 4 above)
Comment (henrayluo) [reply Classes/GlobalStateExplorers/KeyChain/FLEXKeyChainTableViewController.m]: ok

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.m (hunk 11 above)
Comment (henrayluo) [reply Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.m]: Yes, it is from SSKeyChain, and it is [MIT LICENSE](https://github.com/samsoffes/sskeychain/blob/master/LICENSE) and i just renamed the file name and Classes. Should i do something more about the license?

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.h (hunk 10 above)
Comment (henrayluo) [reply Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.h]: it is from [SSKeyChain](https://github.com/soffes/SAMKeychain)

Comment (NSExceptional) [reply Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.h]: @matrush Fixed, testing the UI then I'll push my changes and merge

File: Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.m (hunk 11 above)
Comment (NSExceptional) [reply Classes/GlobalStateExplorers/KeyChain/FLEXKeychain.m]: I will add the license

**Comment by matrush** [comment]
I will leave this to @NSExceptional, thanks for making the changes!

**Comment by NSExceptional** [comment]
Just added a new class to `master`, `FLEXAlert`. Rebased the PR + my work on top of `master` so I can make use of it to create a dialog for adding keychain items. Once I have that done, I'll merge this!

**Pull request merged**